walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,234 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Assume that one person does not have two meetings at the same time
+    
+    //Step 0: Process the raw information
+    int meetingDuration = (int) request.getDuration();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> availableSlots = new ArrayList<TimeRange>();
+
+    // If no attendees, then the whole day is available for the request (assume duration <= 1 day)
+    if (mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty()){
+        availableSlots.add(TimeRange.WHOLE_DAY);
+        return availableSlots;
+        
+    } else { // There are some attendees, either optional or mandatory ones
+        // Step 1: Store the information of the mandatory attenddees provided in events
+
+        // commonCalendar is a lookup table where each person's name is the key
+        // and the associated values are the times in the day when they are busy
+        HashMap<String, ArrayList<TimeRange>> commonCalendar = createCommonCalendar(events);
+
+        // Step 2: Return calendars based on whether there are only mandatory attendees, optional attendees, or both
+        if (mandatoryAttendees.isEmpty() && !optionalAttendees.isEmpty()) {
+
+            ArrayList<TimeRange> availableOptional = getFreeTimeSlots(commonCalendar, optionalAttendees);
+            ArrayList<TimeRange> availableOptionalOnly = TimeRange.getRangesLongEnough(availableOptional, meetingDuration);
+            return TimeRange.getUniqueSortedSlots(availableOptionalOnly);
+
+        } else if (!mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+            ArrayList<TimeRange> allAvailableSlots = getFreeTimeSlots(commonCalendar, mandatoryAttendees);
+            availableSlots = TimeRange.getRangesLongEnough(allAvailableSlots, meetingDuration);
```

Comment:
I think you can merge getRangesLongEnough() into getFreeTimeSlots() by passing the meeting Duration to it, so that you don't need to call two functions together every time

Reply:
I break it up because I don't use getFreeTimeSlots when I'm trying to find the union of the free times of mandatory and optional attendees as I think when I traverse through all of the free slots of the optional people, I can decide whether each of them has a correct duration or not. Essentially I don't want to go through everything two times. Do you think I should still merge the two then and use it for mandatory attendees?